### PR TITLE
Update service metadata in catalog instead of overriding the object during PUT operation

### PR DESCRIPTION
**Pull request by glearner**
### Requirements

*Service api metadata update bug fixed
* All new code requires tests to ensure against regressions

### Description of the Change

In Service api, metadata got replaced by new meta data. I changed the code so that update will consider old data and update it along with old data.

### Benefits

Now consistency of metadata json is maintained.

### Possible Drawbacks

not any in my view

### Applicable Issues

not any in my view

### CODE REVIEW

File: core/jazz_services/components/utils.js [hunk 1]
```diff
@@ -162,7 +163,26 @@ var getUpdateData = function (update_data) {
             } else {
                 input_data[field] = [];
             }
-        } else {
+        }
+        else if (update_data[field] && update_data[field].constructor === Object) {
```
Comment (svidhani) [inline core/jazz_services/components/utils.js]: move this to the line above "} else if{...."

File: core/jazz_services/components/utils.js [hunk 2]
```diff
@@ -144,7 +145,7 @@ var initDynamodb = function () {
 
 //Assigning null to empty string as DynamoDB doesnot allow empty string.
 //But our usecase needs empty string for updation
-var getUpdateData = function (update_data) {
+var getUpdateData = function (update_data,old_data) {
```
Comment (svidhani) [inline core/jazz_services/components/utils.js]: space after comma

File: core/jazz_services/components/utils.js [hunk 3]
```diff
@@ -162,7 +163,26 @@ var getUpdateData = function (update_data) {
...
-        } else {
+        }
+        else if (update_data[field] && update_data[field].constructor === Object) {
+            var ref_data = old_data.validateServiceExists['service_payload'];
+            var new_keys = Object.keys(update_data[field]);
+            var old_keys = Object.keys(ref_data[field]);
+            var update_object = update_data[field];
+            var new_object = ref_data[field];
+            if (new_keys.length > 0) {
+                for (var i = 0; i < new_keys.length; i++) {
```
Comment (svidhani) [inline core/jazz_services/components/utils.js]: use let

File: core/jazz_services/components/utils.js [hunk 4]
```diff
@@ -162,7 +163,26 @@ var getUpdateData = function (update_data) {
...
+                    if(old_keys.indexOf(new_keys[0])>=0){
```
Comment (svidhani) [inline core/jazz_services/components/utils.js]: if (
fix the spacing everywhere else

File: core/jazz_services/components/validation/validate_update_payload.js [hunk 5]
```diff
@@ -118,7 +118,7 @@ module.exports = (service_id, service_data, onComplete) => {
                 onComplete(error);
             } else {
                 logger.info("# Validate Update Payload Data:" + JSON.stringify(data));
-                onComplete(null);
+                onComplete(null,data);
```
Comment (svidhani) [inline core/jazz_services/components/validation/validate_update_payload.js]: spacing

**Comment by svidhani** [comment]
Please fill out the template for this PR.

File: core/jazz_services/components/utils.js (hunk 1 above)
Comment (devsatishm) [inline core/jazz_services/components/utils.js]: Can we add/update unit test(s) to cover this new piece of code?

File: core/jazz_services/components/utils.js (hunk 3 above)
Comment (glearner) [reply core/jazz_services/components/utils.js]: should i have to update whole folder's var or mine only ??

File: core/jazz_services/components/utils.js (hunk 2 above)
Comment (glearner) [reply core/jazz_services/components/utils.js]: done

File: core/jazz_services/components/utils.js (hunk 1 above)
Comment (glearner) [reply core/jazz_services/components/utils.js]: done

File: core/jazz_services/components/utils.js (hunk 3 above)
Comment (glearner) [reply core/jazz_services/components/utils.js]: and i run prettier to fix indentation

**Comment by supritat** [comment]
Comments are incorporated and unit tests are updated.

**Comment by glearner** [comment]
all changes are made please review this code

File: core/jazz_services/components/utils.js [hunk 6]
```diff
@@ -162,7 +164,30 @@ var getUpdateData = function (update_data) {
...
+            update_data[field].constructor === Object &&
+            old_data[field]
+          ) {
+            var ref_data = old_data;
+            var new_keys = Object.keys(update_data[field]);
+            var old_keys = Object.keys(ref_data[field]);
+            var update_object = update_data[field];
+            var new_object = ref_data[field];
+            if (new_keys.length >= 0) {
+              for (var i = 0; i < new_keys.length; i++) {
```
Comment (svidhani) [inline core/jazz_services/components/utils.js]: let?

File: core/jazz_services/test/test.js [hunk 7]
```diff
@@ -685,6 +685,86 @@ describe('platform_services', function() {
...
+  assert.isTrue(logCheck);
+});
+
+/* 
+* Given a successful attempt at a dynamo service update, handler() should indicate array updated successfully
+* @param {object} event -> event.method is defined to be "PUT", event.path.id is defined
+* @params {object, function} default aws context, and callback function as defined in beforeEach
+*/
+it("should indicate that array was updated upon successful documentClient.update()", () => {
+event.method = "PUT";
```
Comment (svidhani) [inline core/jazz_services/test/test.js]: indentation here and other places

**Comment by glearner** [comment]
@svidhani  actually i don't know why let is creating error ... and in whole module only var is used. So can you confirm that it supports ES6 ??

**Comment by svidhani** [comment]
Its because the svc was created using 4.3 and there u have to use 'use strict'

**Pull request merged**